mlflow/store/artifact/databricks_model_artifact_repo.py

```diff
@@ -0,0 +1,144 @@
+import logging
+import json
+
+import mlflow.tracking
+from mlflow.entities import FileInfo
+from mlflow.exceptions import MlflowException
+
+from mlflow.store.artifact.artifact_repo import ArtifactRepository
+from mlflow.utils.databricks_utils import get_databricks_host_creds
+from mlflow.utils.file_utils import download_file_using_signed_uri
+from mlflow.utils.rest_utils import http_request
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+from mlflow.utils.uri import (
+    get_databricks_profile_uri_from_artifact_uri,
+    is_databricks_profile,
+    parse_model_uri,
+    get_model_version_from_stage,
+)
+
+_logger = logging.getLogger(__name__)
+_DOWNLOAD_CHUNK_SIZE = 100000000
+REGISTRY_LIST_ARTIFACTS_ENDPOINT = "/api/2.0/mlflow/model-versions/list-artifacts"
+REGISTRY_ARTIFACT_PRESIGNED_URI_ENDPOINT = "/api/2.0/mlflow/model-versions/get-signed-download-uri"
+
+
+class DatabricksModelArtifactRepository(ArtifactRepository):
+    """
+    Performs storage operations on model registry artifacts in the access-controlled
+    `dbfs:/databricks/model-registry` location
+
+    Signed access URIs for S3 / Azure Blob Storage are fetched from the MLflow service and used to
+    download model artifacts.
+
+    The artifact_uri is expected to be of the form
+    - `models:/<model_name>/<model_version>`
+    - `models:/<model_name>/<stage>`  (refers to the latest model version in the given stage)
+    - `models://<profile>/<model_name>/<model_version or state>`
+
+    Note : This artifact repository is meant is to be instantiated by the ModelsArtifactRepository
+    when the model download uri is of the form
+    `dbfs:/databricks/mlflow-registry/<model-version-id>/models/<artifact-path>`
+    """
+
+    def __init__(self, artifact_uri):
+        if not is_databricks_profile(artifact_uri):
+            raise MlflowException(
+                message="A valid databricks profile is required to use this repository",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        super().__init__(artifact_uri)
+        from mlflow.tracking import MlflowClient
+
+        databricks_profile_uri = (
+            get_databricks_profile_uri_from_artifact_uri(artifact_uri) or mlflow.get_registry_uri()
+        )
+
+        self.client = MlflowClient(registry_uri=databricks_profile_uri)
+        hostcreds_from_uri = get_databricks_host_creds(databricks_profile_uri)
+        self.get_host_creds = lambda: hostcreds_from_uri
+        self.model_name, self.model_version = self._get_model_name_and_version(artifact_uri)
+
+    def _get_model_name_and_version(self, artifact_uri):
+        (model_name, model_version, model_stage) = parse_model_uri(artifact_uri)
+        if model_stage is not None:
+            model_version = get_model_version_from_stage(self.client, model_name, model_stage)
```

Comment:
The way `parse_model_uri` and `get_model_version_from_stage` are used, we could just have a utils function that is equivalent to `_get_model_name_and_version` exposed to DatabricksModelsArtifactRepo and ModelsArtifactRepo. `parse_model_uri` and `get_model_version_from_stage` can be internal functions in the utils module.

This will simplify the constructor here, and we won't need `self.client` (which is also kind of a tricky thing to initialize once, since it's dependent on contextual/environment variables - not quite as relevant here, though).